src/org/zaproxy/zap/spider/parser/SpiderHtmlFormParser.java

```diff
@@ -143,6 +151,7 @@ public boolean parseResource(HttpMessage message, Source source, int depth) {
 			if (method != null && method.trim().equalsIgnoreCase(METHOD_POST)) {
 				// Build the absolute canonical URL
 				String fullURL = URLCanonicalizer.getCanonicalURL(action, baseURL);
+				url = fullURL;
```

Comment:
The `url` should be set before calling `prepareFormDataSet` (the attached test shows that the first `url` passed to form generator is null and for the second is the value of the first form).